File: 07-trees/00199-Binary_Tree_Right_Side_View/00199-binary_tree_right_side_view.cpp
// 199: Binary Tree Right Side View
// https://leetcode.com/problems/binary-tree-right-side-view/


#include <iostream>
#include <queue>
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

TreeNode* creator(vector<int> values, TreeNode** root, int i, int n) {
    if (n==0) return NULL;
    if (i<n) {
        TreeNode *temp = new TreeNode(values[i]);
        *root = temp;
        (*root)->left = creator(values, &((*root)->left), 2*i+1, n);
        (*root)->right = creator(values, &((*root)->right), 2*i+2, n);
    }
    return *root;
};

void createTree(TreeNode** root, vector<int> inputs) {
    creator(inputs, root, 0, inputs.size());
}

void showTree(TreeNode* root) {
    queue<TreeNode*> q;
    vector<vector<int>> result = {};
    vector<int> c;
    if (root==NULL) { cout << "Empty !" << endl; return; }
    q.push(root);
    q.push(NULL);
    while (!q.empty()) {
        TreeNode *t = q.front();
        q.pop();
        if (t==NULL) {
            result.push_back(c);
            c.resize(0);
            if (q.size() > 0) q.push(NULL);
        } else {
            c.push_back(t->val);
            if (t->left) q.push(t->left);
            if (t->right) q.push(t->right);
        }
    }

    cout<<"["; for (auto x : result) {
        cout<<"["; for (auto y : x) {
            if (!y) { cout << "NULL,"; continue; }
            cout<<y<<",";
        } cout<<"\b],";
    } cout<<"\b]"<<endl;
}



class Solution {
private:
    void traverse(TreeNode *root, int level, vector<int> &result) {
        if(root==NULL) return ;
        if(result.size()<level) result.push_back(root->val);
        traverse(root->right, level+1, result);
        traverse(root->left, level+1, result);
    }

public:
    // SOLUTION
    vector<int> rightSideView(TreeNode* root) {
        vector<int> result;
        traverse(root, 1, result);
        return result;
    }
};



int main() {
    Solution o;
    TreeNode* root = NULL;

    // INPUT
    vector<int> tn = {1,2,3,NULL,5,NULL,4};
    createTree(&root, tn);

    // OUTPUT
    auto result = o.rightSideView(root);
    cout<<"["; for (auto y : result) {
        if (!y) { cout << "NULL,"; continue; }
        cout<<y<<",";
    } cout<<"\b]";

    return 0;
}

File: 07-trees/00199-Binary_Tree_Right_Side_View/00199-binary_tree_right_side_view.py
# 199: Binary Tree Right Side View
# https://leetcode.com/problems/binary-tree-right-side-view/

from typing import List

class TreeNode:
    def __init__(self, x: int):
        self.val = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self) -> None:
        self.root = None

    def creator(self, values: list[int], root: TreeNode, i: int, n: int) -> TreeNode:
        if n==0 : return None
        if i<n:
            temp = TreeNode(values[i])
            root = temp
            root.left = self.creator(values, root.left, 2*i+1, n)
            root.right = self.creator(values, root.right, 2*i+2, n)
        return root

    def createTree(self, inputs: list[int]) -> None:
        self.root = self.creator(inputs, self.root, 0, len(inputs))
    
    def showTree(self, root: TreeNode) -> None:
        q = []
        result = []
        c = []
        if root==None : return result
        q.append(root)
        q.append(None)
        while len(q)!=0:
            t = q.pop(0)
            if t==None:
                result.append(c)
                c = []
                if len(q) > 0 : q.append(None)
            else:
                c.append(t.val)
                if t.left!=None : q.append(t.left)
                if t.right!=None : q.append(t.right)

        print("[", end="")
        for x in result:
            if len(x)==0 : continue
            print("[", end="")
            for y in x:
                if y==None: 
                    print("NULL,", end="")
                    continue
                print(y, end=",")
            print("\b]", end=",")
        print("\b]")


class Solution:
    # SOLUTION
    def rightSideView(self, root: TreeNode) -> List[int]:
        result: List[int] = []
        
        def traverse(root: TreeNode, level: int) -> None:
            if root==None: return
            if len(result)<level : result.append(root.val)
            traverse(root.right, level+1)
            traverse(root.left, level+1)

        traverse(root, 1)
        return result


if __name__ == "__main__":
    o = Solution()
    tree = Tree()

    # INPUT
    tn = [1,2,3,None,5,None,4]
    tree.createTree(tn)

    # OUTPUT
    result = o.rightSideView(tree.root)
    tree.showTree(tree.root)
    print(result)
